File: backend/app/models/composition.py
"""组稿 (Composition) 领域模型 —— AST 阶段。

设计要点(与已确认 ADR 对齐):
- 不建 spaces/space_members 表。Folder / Composition 直接持久化 scope_type + owner_id:
  shared → owner_id 为 NULL;personal → owner_id 非 NULL(由 CheckConstraint 保证)。
- Subject 是强上下文:folders / compositions 的 subject_id 均 NOT NULL。
- CompositionNode 取代线性 CompositionBlock,承载一棵组稿 AST:
  * 主键为客户端生成的 String(36) UUID —— 跨 SQLite/MySQL 稳定,免自增/回填。
  * parent_id 自引用(真 FK,ondelete CASCADE):root 节点 parent 为 NULL;module 子节点
    parent 指向所属 module。首期 module 不嵌套(只有 root / module.body 两层)。
  * slot:root 节点为 NULL;module 子节点为 "body"。
  * position:同一 (parent, slot) 下的 0-based 顺序。
  * node_kind:block / module / reference;node_type:细分类型字符串。
  * source_question_node_id / anchor_before_node_id 为“软指针”:只在服务层保证同稿引用
    (不建 DB FK),避免自引用 FK 的 circular DDL 与插入顺序难题。
- Module Definition / Preset 不入库(纯前端/服务层配置)。
- composition_versions 存不可变 snapshot JSON;composition_events 独立于 activity_logs。
"""
import enum
from datetime import datetime

from sqlalchemy import (
    JSON,
    BigInteger,
    Boolean,
    CheckConstraint,
    Column,
    DateTime,
    Enum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    false,
)
from sqlalchemy.orm import relationship

from .base import Base

# BigInteger 供 MySQL 大容量自增 PK;SQLite 仅对 INTEGER PRIMARY KEY 做 rowid 自增,
# 故在 SQLite(桌面版/测试)上退化为 INTEGER 以保留自增语义(仅 events 使用)。
_BigIntPK = BigInteger().with_variant(Integer(), "sqlite")

# 组稿 AST 节点主键:客户端生成的 UUID 文本(str(uuid4()) → 36 字符)。
NODE_ID_LEN = 36

# module 子节点唯一的 slot 名。
BODY_SLOT = "body"


class ScopeType(str, enum.Enum):
    """归属范围:共享(团队可见)或个人私有。"""
    SHARED = "shared"
    PERSONAL = "personal"


class CompositionStatus(str, enum.Enum):
    DRAFT = "draft"
    ARCHIVED = "archived"


class CompositionNodeKind(str, enum.Enum):
    """AST 节点大类。"""
    BLOCK = "block"
    MODULE = "module"
    REFERENCE = "reference"


# node_type 细分(存 String(64),由 schema/service 校验取值合法性)。
NODE_TYPE_RICH_TEXT = "rich_text"
NODE_TYPE_HEADING = "heading"
NODE_TYPE_QUESTION = "question"
NODE_TYPE_PAGE_BREAK = "page_break"
NODE_TYPE_ANSWER_SPACE = "answer_space"
NODE_TYPE_QUESTION_DETAILS = "question_details"
NODE_TYPE_ANSWER_ITEM = "answer_item"

# 各 kind 允许的 node_type 集合(服务层/契约共享)。
BLOCK_NODE_TYPES = frozenset(
    {
        NODE_TYPE_RICH_TEXT,
        NODE_TYPE_HEADING,
        NODE_TYPE_QUESTION,
        NODE_TYPE_PAGE_BREAK,
        NODE_TYPE_ANSWER_SPACE,
    }
)
MODULE_NODE_TYPES = frozenset({NODE_TYPE_QUESTION_DETAILS})
REFERENCE_NODE_TYPES = frozenset({NODE_TYPE_ANSWER_ITEM})


def _enum_col(enum_cls: type[enum.Enum]) -> Enum:
    """按仓库惯例:以 value(而非 name)持久化枚举。"""
    return Enum(enum_cls, values_callable=lambda obj: [e.value for e in obj])


class Folder(Base):
    """组稿文件夹:支持父子层级,承载 scope/owner/subject 上下文。"""
    __tablename__ = "folders"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)

    scope_type = Column(_enum_col(ScopeType), nullable=False)
    # shared → NULL;personal → 非 NULL(由 ck_folders_scope_owner 保证)。
    owner_id = Column(Integer, ForeignKey("user.id"), nullable=True, index=True)
    # Subject 为强上下文。
    subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=False, index=True)

    parent_id = Column(Integer, ForeignKey("folders.id"), nullable=True, index=True)

    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(Integer, ForeignKey("user.id"), nullable=True)
    updated_by = Column(Integer, ForeignKey("user.id"), nullable=True)
    deleted_at = Column(DateTime, nullable=True, index=True)

    owner = relationship("User", foreign_keys=[owner_id])
    creator = relationship("User", foreign_keys=[created_by])
    updater = relationship("User", foreign_keys=[updated_by])
    subject = relationship("Subject")
    children = relationship(
        "Folder",
        backref="parent",
        remote_side=[id],
        # 单向:此处仅声明父子导航,子->父由 backref 提供。
    )
    compositions = relationship("Composition", back_populates="folder")

    __table_args__ = (
        CheckConstraint(
            "(scope_type = 'shared' AND owner_id IS NULL) OR "
            "(scope_type = 'personal' AND owner_id IS NOT NULL)",
            name="scope_owner",
        ),
        Index("ix_folders_subject_scope_owner", "subject_id", "scope_type", "owner_id"),
        # 注意:父子文件夹之间 subject/scope/owner 必须一致,但该不变量跨行、无法用简单
        # 行级 CheckConstraint 表达。此处不伪造数据库保证,改由服务层强制。
    )


class Composition(Base):
    """组稿:一份可编辑的稿件,由一棵节点 AST 组成,可产生不可变版本快照。"""
    __tablename__ = "compositions"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    status = Column(String(20), default=CompositionStatus.DRAFT.value, nullable=False)
    revision = Column(Integer, default=1, nullable=False)
    # 题号开关:关闭仅隐藏显示,题号仍保留在 question 节点 props.number 上。
    numbering_enabled = Column(Boolean, nullable=False, default=False, server_default=false())
    # 赋分开关:仅在 numbering_enabled 为真时可开启;关闭仅隐藏显示,分值仍保留在 question 节点 props.score 上。
    scoring_enabled = Column(Boolean, nullable=False, default=False, server_default=false())
    # 画布题目全局显示字段(answer/thinking/analysis/summary);NULL 视为全部隐藏。
    question_display = Column(JSON, nullable=True)

    scope_type = Column(_enum_col(ScopeType), nullable=False)
    owner_id = Column(Integer, ForeignKey("user.id"), nullable=True, index=True)
    subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=False, index=True)
    folder_id = Column(Integer, ForeignKey("folders.id"), nullable=True, index=True)

    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(Integer, ForeignKey("user.id"), nullable=True)
    updated_by = Column(Integer, ForeignKey("user.id"), nullable=True)
    deleted_at = Column(DateTime, nullable=True, index=True)

    owner = relationship("User", foreign_keys=[owner_id])
    creator = relationship("User", foreign_keys=[created_by])
    updater = relationship("User", foreign_keys=[updated_by])
    subject = relationship("Subject")
    folder = relationship("Folder", back_populates="compositions")

    # passive_deletes:节点删除依赖 DB 的 ON DELETE CASCADE(含自引用),ORM 不逐行删,
    # 避免自引用父子删除顺序问题(需 SQLite PRAGMA foreign_keys=ON 生效)。
    nodes = relationship(
        "CompositionNode",
        back_populates="composition",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="CompositionNode.position",
    )
    versions = relationship(
        "CompositionVersion",
        back_populates="composition",
        cascade="all, delete-orphan",
        order_by="CompositionVersion.version_no",
    )
    events = relationship(
        "CompositionEvent",
        back_populates="composition",
        cascade="all, delete-orphan",
        order_by="CompositionEvent.created_at",
    )

    __table_args__ = (
        CheckConstraint(
            "(scope_type = 'shared' AND owner_id IS NULL) OR "
            "(scope_type = 'personal' AND owner_id IS NOT NULL)",
            name="scope_owner",
        ),
        Index("ix_compositions_subject_scope_owner", "subject_id", "scope_type", "owner_id"),
    )


class CompositionNode(Base):
    """组稿 AST 的一个节点(block / module / reference)。"""
    __tablename__ = "composition_nodes"

    # 客户端生成的 UUID 文本主键。
    id = Column(String(NODE_ID_LEN), primary_key=True)
    composition_id = Column(
        Integer, ForeignKey("compositions.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # 自引用父指针(真 FK,ondelete CASCADE):root 节点为 NULL。
    parent_id = Column(
        String(NODE_ID_LEN),
        ForeignKey("composition_nodes.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    slot = Column(String(32), nullable=True)  # root=NULL;module 子节点="body"
    position = Column(Integer, nullable=False)  # 同 (parent, slot) 内 0-based 顺序

    node_kind = Column(_enum_col(CompositionNodeKind), nullable=False)
    node_type = Column(String(64), nullable=False)

    # 结构化内容:rich_text/heading 存 RichDoc;question 存冻结的题目内容快照;其余为 NULL。
    content = Column(JSON, nullable=True)
    props = Column(JSON, nullable=True)
    schema_version = Column(Integer, default=1, nullable=False)

    # 仅 question 节点使用:引用题目及被钉住的修订号。
    question_id = Column(Integer, ForeignKey("questions.id"), nullable=True, index=True)
    question_revision = Column(Integer, nullable=True)

    # 软指针(无 DB FK,服务层保证同稿):
    # - reference.answer_item.source_question_node_id → 同稿 root 层 question 节点。
    # - 自定义 module 子节点.anchor_before_node_id → 同 module 内 answer_item 节点。
    source_question_node_id = Column(String(NODE_ID_LEN), nullable=True)
    anchor_before_node_id = Column(String(NODE_ID_LEN), nullable=True)

    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    created_by = Column(Integer, ForeignKey("user.id"), nullable=True)
    updated_by = Column(Integer, ForeignKey("user.id"), nullable=True)

    composition = relationship("Composition", back_populates="nodes")
    question = relationship("Question")
    creator = relationship("User", foreign_keys=[created_by])
    updater = relationship("User", foreign_keys=[updated_by])

    __table_args__ = (
        CheckConstraint("position >= 0", name="position_non_negative"),
        CheckConstraint(
            "(parent_id IS NULL AND slot IS NULL) OR "
            "(parent_id IS NOT NULL AND slot = 'body')",
            name="parent_slot_matches",
        ),
        CheckConstraint(
            "(node_kind = 'block' AND node_type IN "
            "('rich_text', 'heading', 'question', 'page_break', 'answer_space')) OR "
            "(node_kind = 'module' AND node_type = 'question_details') OR "
            "(node_kind = 'reference' AND node_type = 'answer_item')",
            name="kind_matches_type",
        ),
        # question 节点必须携带完整题目引用与冻结快照;其他节点不得携带。
        CheckConstraint(
            "(node_type = 'question' AND question_id IS NOT NULL AND "
            "question_revision IS NOT NULL AND content IS NOT NULL) OR "
            "(node_type <> 'question' AND question_id IS NULL AND "
            "question_revision IS NULL)",
            name="question_ref_matches_type",
        ),
        CheckConstraint(
            "(node_type = 'answer_item' AND source_question_node_id IS NOT NULL "
            "AND parent_id IS NOT NULL) OR "
            "(node_type <> 'answer_item' AND source_question_node_id IS NULL)",
            name="source_ref_matches_type",
        ),
        CheckConstraint(
            "anchor_before_node_id IS NULL OR "
            "(parent_id IS NOT NULL AND node_type IN ('heading', 'rich_text'))",
            name="anchor_matches_type",
        ),
        Index(
            "ix_composition_nodes_comp_parent_slot_pos",
            "composition_id", "parent_id", "slot", "position",
        ),
    )


class CompositionVersion(Base):
    """组稿的不可变版本快照。"""
    __tablename__ = "composition_versions"

    id = Column(Integer, primary_key=True, index=True)
    composition_id = Column(
        Integer, ForeignKey("compositions.id", ondelete="CASCADE"), nullable=False, index=True
    )
    version_no = Column(Integer, nullable=False)  # 每个 composition 内单调递增
    source_revision = Column(Integer, nullable=False)
    title = Column(String(255), nullable=False)
    subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=False)
    snapshot = Column(JSON, nullable=False)  # 不可变的完整快照
    label = Column(String(255), nullable=True)

    finalized_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    finalized_by = Column(Integer, ForeignKey("user.id"), nullable=False)

    composition = relationship("Composition", back_populates="versions")
    subject = relationship("Subject")
    finalizer = relationship("User", foreign_keys=[finalized_by])

    __table_args__ = (
        UniqueConstraint("composition_id", "version_no", name="composition_version_no"),
    )


class CompositionEvent(Base):
    """组稿时间线事件,独立于 activity_logs(系统审计)。"""
    __tablename__ = "composition_events"

    id = Column(_BigIntPK, primary_key=True)
    composition_id = Column(
        Integer, ForeignKey("compositions.id", ondelete="CASCADE"), nullable=False, index=True
    )
    composition_revision = Column(Integer, nullable=False)
    event_type = Column(String(50), nullable=False)
    target_type = Column(String(50), nullable=True)
    target_id = Column(String(64), nullable=True)
    summary = Column(String(255), nullable=False)
    payload = Column(JSON, nullable=True)
    batch_id = Column(String(64), nullable=True)

    actor_id = Column(Integer, ForeignKey("user.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    composition = relationship("Composition", back_populates="events")
    actor = relationship("User", foreign_keys=[actor_id])

    __table_args__ = (
        Index("ix_composition_events_comp_id", "composition_id", "id"),
        Index("ix_composition_events_actor_created", "actor_id", "created_at"),
    )

File: frontend/app/types/composition.ts
// 组稿 (Composition) 领域类型 —— CompositionNode AST 阶段。
// 与后端 app/schemas/composition.py 严格对齐（node 契约 + snapshot v2）。

import type { RichDocNode } from './richContent'
import type { AnswerSpec, OptionSpec, QuestionType } from './question'

export type CompositionScope = 'shared' | 'personal'

export type CompositionStatus = 'draft' | 'archived'

export interface CompositionFolder {
  id: number
  name: string
  scope_type: CompositionScope
  owner_id: number | null
  subject_id: number
  parent_id: number | null
  created_at: string
  updated_at: string
}

export interface Composition {
  id: number
  title: string
  description: string | null
  status: CompositionStatus
  revision: number
  numbering_enabled: boolean
  scoring_enabled: boolean
  question_display: Record<AnswerFieldKey, boolean>
  scope_type: CompositionScope
  owner_id: number | null
  subject_id: number
  folder_id: number | null
  created_at: string
  updated_at: string
  deleted_at: string | null
}

// --- 请求体（scope/subject/owner 由路径与鉴权强制，前端绝不传递 owner） --- //

export interface FolderCreatePayload {
  name: string
  parent_id?: number | null
}

export interface FolderUpdatePayload {
  name?: string
  // 显式传 null 表示移动到根目录；后端用 model_fields_set 区分未提供与置空。
  parent_id?: number | null
}

export interface CompositionCreatePayload {
  title: string
  description?: string | null
  folder_id?: number | null
}

export interface CompositionMetaUpdatePayload {
  expected_revision: number
  title?: string
  description?: string | null
  status?: CompositionStatus
  // 显式传 null 表示移出目录（根）。
  folder_id?: number | null
  numbering_enabled?: boolean
  scoring_enabled?: boolean
  question_display?: Record<AnswerFieldKey, boolean>
}

// 折叠成树后的目录节点。
export interface CompositionFolderNode extends CompositionFolder {
  children: CompositionFolderNode[]
}

// --- Node 判别联合（与后端 CompositionNode* 对齐） --- //

export type CompositionNodeKind = 'block' | 'module' | 'reference'

export type CompositionNodeType =
  | 'rich_text'
  | 'heading'
  | 'question'
  | 'page_break'
  | 'answer_space'
  | 'question_details'
  | 'answer_item'

// module 子节点唯一 slot 名（与后端 BODY_SLOT 对齐）。
export const BODY_SLOT = 'body'

export type HeadingLevel = 1 | 2 | 3 | 4

export interface HeadingProps {
  level: HeadingLevel
}

/** 作答空间样式：纯空白或答题横线。 */
export type AnswerSpaceStyle = 'blank' | 'lined'

/** answer_space 节点属性：按行数控高 + 样式。 */
export interface AnswerSpaceProps {
  lines: number
  style: AnswerSpaceStyle
}

/** 选项排版：auto=按内容长度自适应列数，或固定 1/2/4 列。缺省视为 auto。 */
export type OptionLayout = 'auto' | 1 | 2 | 4

/** question 节点可选题号（纠数字或 1.1 形式），存于 node.props。 */
export interface QuestionProps {
  number?: string | null
  // 题目级字段显隐覆盖：true=显示、false=隐藏、缺省/null=继承全局。
  show?: Partial<Record<AnswerFieldKey, boolean | null>>
  // 选项排版覆盖（每卷生效）；缺省/'auto' 由内容长度自适应。
  optionLayout?: OptionLayout
  // 题目分值（0~1000，允许 0.5 步进小数）；仅在组稿 scoring_enabled 为真时展示/可编辑。
  score?: number | null
}

// question_details / answer_item 覆盖涉及的四个可发布字段。
export type AnswerFieldKey = 'answer' | 'thinking' | 'analysis' | 'summary'

export const ANSWER_FIELD_KEYS: readonly AnswerFieldKey[] = [
  'answer',
  'thinking',
  'analysis',
  'summary',
]

export type DetailScope = 'before' | 'all'

/** question_details module 属性：范围 + 四字段全局开关。 */
export interface QuestionDetailsProps {
  scope: DetailScope
  fields: Record<AnswerFieldKey, boolean>
}

/** answer_item 覆盖：null=继承 module 全局开关，true/false=显式显示/隐藏。 */
export type AnswerItemOverride = boolean | null

export interface AnswerItemProps {
  included: boolean
  overrides: Record<AnswerFieldKey, AnswerItemOverride>
}

/**
 * 冻结在 question 节点上的题目内容快照（不含 id/revision）。
 * 与后端 QuestionContentSnapshot 严格对齐；id 由 question_id、revision 由 question_revision 承载。
 */
export interface QuestionContentSnapshot {
  content_schema_version: number
  q_type: QuestionType
  content: RichDocNode | null
  options: OptionSpec[] | null
  answer: AnswerSpec | null
  thinking: RichDocNode | null
  analysis: RichDocNode | null
  summary: RichDocNode | null
  difficulty: number
  source: string | null
}

interface CompositionNodeCommon {
  id: string
  composition_id: number
  parent_id: string | null
  slot: string | null
  position: number
  schema_version: number
}

export interface RichTextNode extends CompositionNodeCommon {
  node_kind: 'block'
  node_type: 'rich_text'
  content: RichDocNode
  props: null
  question_id: null
  question_revision: null
  source_question_node_id: null
  anchor_before_node_id: string | null
}

export interface HeadingNode extends CompositionNodeCommon {
  node_kind: 'block'
  node_type: 'heading'
  content: RichDocNode
  props: HeadingProps
  question_id: null
  question_revision: null
  source_question_node_id: null
  anchor_before_node_id: string | null
}

export interface QuestionNode extends CompositionNodeCommon {
  node_kind: 'block'
  node_type: 'question'
  content: QuestionContentSnapshot | null
  props: null
  question_id: number
  question_revision: number
  source_question_node_id: null
  anchor_before_node_id: null
}

export interface PageBreakNode extends CompositionNodeCommon {
  node_kind: 'block'
  node_type: 'page_break'
  content: null
  props: null
  question_id: null
  question_revision: null
  source_question_node_id: null
  anchor_before_node_id: null
}

export interface AnswerSpaceNode extends CompositionNodeCommon {
  node_kind: 'block'
  node_type: 'answer_space'
  content: null
  props: AnswerSpaceProps
  question_id: null
  question_revision: null
  source_question_node_id: null
  anchor_before_node_id: null
}

export interface QuestionDetailsNode extends CompositionNodeCommon {
  node_kind: 'module'
  node_type: 'question_details'
  content: null
  props: QuestionDetailsProps
  question_id: null
  question_revision: null
  source_question_node_id: null
  anchor_before_node_id: null
}

export interface AnswerItemNode extends CompositionNodeCommon {
  node_kind: 'reference'
  node_type: 'answer_item'
  content: null
  props: AnswerItemProps
  question_id: null
  question_revision: null
  source_question_node_id: string
  anchor_before_node_id: null
}

/** 服务端返回的 node 判别联合（GET detail / replace / sync 响应共用）。 */
export type CompositionNode =
  | RichTextNode
  | HeadingNode
  | QuestionNode
  | PageBreakNode
  | AnswerSpaceNode
  | QuestionDetailsNode
  | AnswerItemNode

/** GET composition detail：元数据 + 有序 nodes（含 module 子节点）。 */
export interface CompositionDetail extends Composition {
  nodes: CompositionNode[]
}

// --- 整体替换请求/响应契约（AST 契约） --- //
// 每个 node 携带客户端生成的 UUID id；position 不由客户端传入，由服务端按 (parent, slot) 顺序规范化。
// question 节点的 content / question_revision 由服务端冻结/钉住，客户端传值被忽略。

export interface CompositionNodeInput {
  id: string
  parent_id?: string | null
  slot?: string | null
  node_kind: CompositionNodeKind
  node_type: CompositionNodeType
  content?: RichDocNode | null
  props?: Record<string, unknown> | null
  schema_version?: number
  question_id?: number | null
  source_question_node_id?: string | null
  anchor_before_node_id?: string | null
}

export interface CompositionNodesReplaceRequest {
  expected_revision: number
  batch_id?: string
  nodes: CompositionNodeInput[]
}

export interface CompositionNodesReplaceResponse {
  revision: number
  nodes: CompositionNode[]
}

// --- Question node 版本状态 / 同步契约 --- //

/** 稿件内某 question_id 的实时状态（仅状态，不含题目内容）。 */
export interface QuestionRevisionStatus {
  question_id: number
  // available=false（软删/缺失）时为 null。
  current_revision: number | null
  available: boolean
}

/** 同步请求：expected_revision 乐观校验；node_ids 唯一非空（"同步此题"传一个，"同步全部"传全部）。 */
export interface CompositionQuestionNodesSyncRequest {
  expected_revision: number
  node_ids: string[]
}

/** 同步响应：自增后的 revision + 刷新后的完整 node 序列。 */
export interface CompositionQuestionNodesSyncResponse {
  revision: number
  nodes: CompositionNode[]
}

// --- 定稿 (Version) 契约 --- //

/** 定稿请求体：expected_revision 乐观校验（不修改 revision），label 可选备注。 */
export interface CompositionVersionCreatePayload {
  expected_revision: number
  label?: string | null
}

/** 版本导出格式：与后端 app.schemas.paper.OutputFormat 对齐。 */
export type CompositionExportFormat = 'docx' | 'latex'

/** 版本导出请求体：title 可选覆盖导出文件里的标题。 */
export interface CompositionExportPayload {
  format: CompositionExportFormat
  title?: string
}

/** 版本列表项：不含 snapshot，避免列表响应携带整稿快照。 */
export interface CompositionVersionSummary {
  id: number
  composition_id: number
  version_no: number
  source_revision: number
  title: string
  subject_id: number
  label: string | null
  finalized_at: string
  finalized_by: number
}

// --- 时间线 (Event) 契约 --- //

/** 与后端 CompositionEvent.event_type 对齐的已知取值；后端仍可能新增，故不做穷尽联合校验。 */
export type CompositionEventType =
  | 'created'
  | 'updated'
  | 'moved'
  | 'deleted'
  | 'restored'
  | 'nodes_replaced'
  | 'question_nodes_synced'
  | 'finalized'

export interface CompositionEventActor {
  id: number
  username: string
  full_name?: string | null
}

export interface CompositionEvent {
  id: number
  composition_id: number
  composition_revision: number
  event_type: CompositionEventType
  target_type: string | null
  target_id: string | null
  summary: string
  payload: Record<string, unknown> | null
  batch_id: string | null
  actor_id: number
  actor: CompositionEventActor | null
  created_at: string
}

/** 时间线游标分页响应：has_more=true 时可用最后一条 id 作为下一页 before_id。 */
export interface CompositionEventPage {
  items: CompositionEvent[]
  has_more: boolean
}


/**
 * 定稿时冻结的题目内容投影（snapshot v2 内 question 节点携带）。
 * 只含可发布内容，排除关系（标签/知识点/创建人）、权限与审核字段。
 */
export interface QuestionSnapshot {
  id: number
  content_revision: number
  content_schema_version: number
  q_type: QuestionType
  content: RichDocNode | null
  options: OptionSpec[] | null
  answer: AnswerSpec | null
  thinking: RichDocNode | null
  analysis: RichDocNode | null
  summary: RichDocNode | null
  difficulty: number
  source: string | null
}

// --- snapshot v2 node 判别联合（只读，仅消费快照，不查询当前题库） --- //

interface SnapshotNodeCommon {
  id: string
  parent_id: string | null
  slot: string | null
  position: number
  schema_version: number
  anchor_before_node_id?: string
}

export interface SnapshotRichTextNode extends SnapshotNodeCommon {
  node_kind: 'block'
  node_type: 'rich_text'
  content: RichDocNode
}

export interface SnapshotHeadingNode extends SnapshotNodeCommon {
  node_kind: 'block'
  node_type: 'heading'
  content: RichDocNode
  props: HeadingProps
}

export interface SnapshotQuestionNode extends SnapshotNodeCommon {
  node_kind: 'block'
  node_type: 'question'
  question_id: number
  question_revision: number
  question: QuestionSnapshot
  // 冻结的表现层覆盖（选项排版等）；旧快照无此字段时按缺省处理。
  props?: QuestionProps | null
}

export interface SnapshotPageBreakNode extends SnapshotNodeCommon {
  node_kind: 'block'
  node_type: 'page_break'
}

export interface SnapshotAnswerSpaceNode extends SnapshotNodeCommon {
  node_kind: 'block'
  node_type: 'answer_space'
  props: AnswerSpaceProps
}

export interface SnapshotQuestionDetailsNode extends SnapshotNodeCommon {
  node_kind: 'module'
  node_type: 'question_details'
  props: QuestionDetailsProps
}

export interface SnapshotAnswerItemNode extends SnapshotNodeCommon {
  node_kind: 'reference'
  node_type: 'answer_item'
  source_question_node_id: string
  props: AnswerItemProps
}

export type SnapshotNode =
  | SnapshotRichTextNode
  | SnapshotHeadingNode
  | SnapshotQuestionNode
  | SnapshotPageBreakNode
  | SnapshotAnswerSpaceNode
  | SnapshotQuestionDetailsNode
  | SnapshotAnswerItemNode

/** snapshot v2：顶层元数据 + 前序展平的规范化节点（root 按 position，module 子节点紧随其后）。 */
export interface CompositionSnapshotV2 {
  schema_version: 2
  composition_id: number
  source_revision: number
  title: string
  subject_id: number
  finalized_at: string
  // 定稿时刻冻结的显示默认值；旧快照（功能上线前定稿）没有这三个字段，消费方需按 false/全 false 兜底。
  numbering_enabled?: boolean
  scoring_enabled?: boolean
  question_display?: Record<AnswerFieldKey, boolean>
  nodes: SnapshotNode[]
}

/** GET version detail：版本元数据 + 不可变 snapshot。 */
export interface CompositionVersionDetail {
  id: number
  composition_id: number
  version_no: number
  source_revision: number
  title: string
  subject_id: number
  snapshot: CompositionSnapshotV2
  label: string | null
  finalized_at: string
  finalized_by: number
}
